Rewrite (#2)
mihaibudiu/sigmod-rh: intro.tex

\section{Introduction}\label{sec:introduction}

Incremental view maintenance (IVM) is an important and well-studied
problem in databases~\cite{gupta-idb95}.  The IVM problem can be
stated as follows: we are given a large database $DB$ (say 1 billion
records) and a view $V$, described by a query $Q$.  The goal of IVM is
to keep the contents of $V$ up-to-date in response to changes of the
database.

As a concrete example, consider the following view definition
statement in SQL:

\texttt{CREATE VIEW V AS SELECT * FROM T WHERE Age >= 10}.

In this example the query $Q$ defining the view $V$ is
\\ \texttt{SELECT * FROM T WHERE Age >= 10}.  The view \code{V} always
contains all the rows of table \code{T} whose value for the column

> \code{Age} is greater or equal to 10.

Reviewer: This last sentence can be removed if we need to save space.

%mathematical properties --- specifically, that it forms a commutative
%group.  (Fortunately, relational databases can be modeled in such a
%way~\cite{green-pods07, koch-pods10}.)

The \dbsp language has only 4 operators.  However, it can express a
rich set of computations on streams, including repeated computations
(similar to the repeated queries $Q$ above), recursive computations
that compute fixed points (as done by Datalog programs), streaming
computations, and incremental computations (which we define shortly).
The full paper~\cite{budiu-vldb23} gives a precise mathematical
description of \dbsp, this presentation is simplified to convey the
main intuitions behind the design.  We omit the related work section
from this presentation.

The central result of this paper is algorithm~\ref{algorithm-inc}
which, given a \dbsp program that computes on a stream of values,
mechanically transforms it into an incremental \dbsp program that
computes on a stream of changes.

\dbsp is not tied to databases in any way; it is in fact a
Turing-complete language that can be used for many other purposes.
But it works particularly well in the area of databases, for two
reasons:

\begin{itemize}

>   \item \dbsp operates on values that must have nice mathematical
>     properties (it requires the values in a stream to belong to a
>     commutative group).  Databases can be modeled as values from a
>     commutative group.

Reviewer: ```suggestion
  \item \dbsp operates on values that must form a commutative group.  Databases can be modeled as values from a
    commutative group.
```

databases, a notion that we will soon make more precise.  The $\Delta
DB$ stream can be produced from the $DB$ stream by the \emph{stream
differentiation} operator $\D$; this operator produces as its output
the stream of changes from its input stream; we have thus $\D(DB) =
\Delta DB$.

Conversely, the database snapshot at time $t$ is the cumulative result
of applying all transactions up to $t$: $DB[t] = \Delta DB[0] + \Delta
DB[1] + \ldots + \Delta DB[t]$.  The stream operator $\I$ is defined
to produce an output by adding up all previous inputs.  We call this
operator \emph{stream integration}, the inverse of differentiation.
The following diagram shows the relationship between the streams
$\Delta DB$ and $DB$:
\begin{center}
\begin{tikzpicture}[auto,>=latex,minimum width=.5cm, node distance=1.2cm]
  \node[] (input) {$\Delta DB$};
  \node[block, right of=input] (I) {$\I$};
  \node[right of=I] (output) {$DB$};
  \node[block, right of=output] (D) {$\D$};
  \node[right of=D] (end) {$\Delta DB$};
  \draw[->>] (input) -- (I);
  \draw[->>] (I) -- (output);
  \draw[->>] (output) -- (D);
  \draw[->>] (D) -- (end);
\end{tikzpicture}
\end{center}

> Suppose we are given a query $Q$ defining a view $V$.  What is a view
> in our streaming model?  It is also a stream!  For each snapshot of
> the database stream we have a snapshot of the view: $V[t] = Q(DB[t])$.
> In general, given an arbitrary function $f: A \to B$, we define a
> streaming ``version'' of $f$, denoted by $\lift{f}$ (read as ``$f$
> lifted''), which applies $f$ to every element of the input stream
> independently.  We can thus write $V = (\lift{Q})(DB)$.

Reviewer: ```suggestion
What is a view in our streaming model?  It is also a stream!  
Suppose we are given a query $Q$ defining a view $V$.  
For each snapshot of
the database stream we have a snapshot of the view: $V[t] = Q(DB[t])$.
In general, given an arbitrary function $f: A \to B$, we define a
streaming ``version'' of $f$, denoted by $\lift{f}$ (read as ``$f$
lifted''), which applies $f$ to every element of the input stream
independently.  We can thus write $V = (\lift{Q})(DB)$.
```
